### ISSUE TITLE
Syntax highlighting improvements

### ISSUE THREAD

**Issue by gchp** [gchp/iota#127]
The current syntax highlighting solution uses a simple lexer to generate a stream of tokens. Adding new languages means adding a new implementation of the lexer to add behaviour for that specific language. Color values are also hard coded right now.

I'd like to look into a better solution. Regular expressions is one way this could go ([syntect](https://github.com/trishume/syntect) could be interesting). Another way would be to implement some kind of AST based parser which would allow for much more granular & meaningful highlighting, I think.

Basically, I'm not sure what the best approach is, but I don't think the way it is now is it. Current solution works, but I want something better.

Comments in #88 will be useful in this.

**Comment by abaez** [comment]
Hello, 

I was researching a while ago, in making the implementation for issue #88. Just never took the time to actually do it. One thing that I did spent time in was researching and comparing implementations of grammar parsers for syntax highlighting. 

Now it could be because I'm bias to [textadept's way][1], but after looking  over how [vim] and [sublime text][sublime] have gone about it, I've come to the realization that using an AST parser defined by a PEG generator would be best. Right now there's two generators written in rust, which come to mind for the process; [rusty-peg] and [oak]. These two would allow to get away with making a lexer definition which is highly expressive, with leaner symantics, and less of a resource hog for the editor at hand.

The reason why I bring this up is for the ease in definitions of a lexer for Iota and the ease in implementing the parser should both be simple enough to refactor and modify by contributors. Using a PEG generator would allow [_very_][4] [_easy_][5] [definitions][6] [of][7] [lexers][8]. And I've noticed that PEG generated lexers, like [scintillua][9], can be less daunting for the editor at hand. 

In any case, here's an [old talk about the power of PEG generator][10], which is still relevant to the issue at hand and can probably speak better on the advantage over regex parser. 

[1]: http://foicica.com/textadept/api.html#lexer

[4]: https://github.com/abaez/ta-rust/blob/master/rust.lua
[5]: https://github.com/abaez/ta-toml/blob/master/toml.lua
[6]: http://foicica.com/hg/scintillua/file/8f9504cd013e/lexers/markdown.lua
[7]: http://foicica.com/hg/scintillua/file/8f9504cd013e/lexers/go.lua
[8]: http://foicica.com/hg/scintillua/file/8f9504cd013e/lexers/ansi_c.lua

[9]: http://foicica.com/hg/scintillua
[10]: https://vimeo.com/1485123

[vim]: https://github.com/vim/vim/blob/master/runtime/syntax/javascript.vim
[sublime]: https://github.com/sublimehq/Packages/blob/master/JavaScript/JavaScript.sublime-syntax
[rusty-peg]: https://github.com/kevinmehall/rust-peg
[oak]: http://hyc.io/rust-lib/oak/index.html

**Comment by luciusmagn** [comment]
Have you seen the way syntax highlighting works in Nano? Extremely simple but quite effective. With a few improvements, it could work nicely. It is regex based
